Added support for StepArgumentTransformation methods which have multi…

samholder opened the pull request:
This pull request adds support for [StepArgumentTransformation] methods which have more than 1 argument. This allows a step argument transformations to construct more complex objects than a single argument can hold, and prevents you having to munge multiple bits of data into a single string.

Simple example:

imagine this step:
    
     Given I have a time range from 10-Dec-2015 to 11-Dec-2015

traditionally this might be mapped to a step method like:

    [Given("I have a time range from (.*) to (.*)")]
    public void GivenIHaveATimeRange(DateTime start, DateTime end)
    {
     ...
    }

With this PR the following is now possible

    [Given("I have a time range from (.*) to (.*)")]
    public void GivenIHaveATimeRange(TimeRange timeRange)
    {
     ...
    }   

    [StepArgumentTransformation]
    public TimeRange TransformToTimeRange(DateTime start, DateTime end)
    {
          return new TimeRange(start,end);
    }

More complex example:

It is also possible to combine with tables to allow you to create objects which have some state and a collection

imagine this step:
    
     Given I have a user named 'Bob' aged 35 with the following children:
          | childName| childAge | naughtyness|
          | Tom          |  7   |   3        |
          | Carol        |  9   |   1        |

This mapping can be done like this

    [Given("Given I have a user named '(.*)' aged (.*) with the following children:")]
    public void GivenIHaveAUser(User user)
    {
     ...
    }   

    [StepArgumentTransformation]
    public User TransformToUser(string name, int age, Table children)
    {
          ...Create user from all parameters and return
    }

Multiple step argument transformations are supported so you could combine the above 2 examples with a simple string value in a method definition like this:

    [Given("I have a user who was employed between (.*) and (.*) at the company '(.*)'. The user is named '(.*)' aged (.*) with the following children:")]
    public void GivenIHaveAUser(TimeRange employmentRange, string companyName, User user)
    {
     ...
    }   

The restriction is that step parameters will be consumed in the given order to create the step argument transformation, so anything which consumes a table must use last parameters in the list.

darrencauthon commented:
:+1:

samholder commented:
please hold off from merging this as I am investigating merging it with #397 which has better code structure and some other improvements

samholder commented:
If the CI build passes then this should be ok to merge and should include the improvements in #397 as well. any code review appreciated :smile:

darrencauthon commented:
I'll take a deeper look tonight 



—
Sent from Mailbox

On Thu, Jul 9, 2015 at 4:32 PM, Sam Holder <notifications@github.com>
wrote:

> If the CI build passes then this should be ok to merge and should include the improvements in #397 as well. any code review appreciated :smile:
>  
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/techtalk/SpecFlow/pull/451#issuecomment-120149737

darrencauthon commented on Runtime/Bindings/StepArgumentTypeConverter.cs:
```diff
@@ -1,151 +1,42 @@
+                new IndentityConverter(),
+                new StepArgumentTransformationConverter(this, testTracer, bindingRegistry, contextManager, bindingInvoker),
+                new VerticalTableConverter(this),
+                new HorizontalTableConverter(this),
```
One thing I noticed while poking around the code & tests is that these two lines are not tested.  I can comment ```new VerticalTable..``` and ```new HorizontalTable...``` but the tests still pass.

I thought it might just be an integration issue, but I was also able to remove the code inside ```CanConvert``` and ```Convert``` from both classes and still was able to get the tests to pass.

samholder replied:
Those changes came from the #397 fork, which was from the master branch. I think the code had been refactored to place the transforms into separate classes for simple, StepArgumentTypeConverter, tables etc but the v2 branch was taken from before that had happened. It was easier for me to take those changes and redo my changes over the top than it was to try and add the changes from #397 into my changes due to the changes in the structure and my changes only needed to change the StepArgumentTypeConverter (apart from the signature changes).

darrencauthon commented on Runtime/Bindings/VerticalTableConverter.cs:
```diff
@@ -0,0 +1,149 @@
+            this.stepArgumentTypeConverter = stepArgumentTypeConverter;
+        }
+
+        protected override object Convert(Table table, IBindingType typeToConvertTo, CultureInfo cultureInfo)
```
If there was a non-generic method override for ```CreateInstance``` and ```CreateSet```, could this logic rely on the Assist methods that create objects from tables?

samholder replied:
Possibly. I'll try and look at this again with that in mind over the weekend, if I get enough time free from painting the house.

darrencauthon commented:
These are minor details, but I came across some more code that appears to be untested:

https://github.com/darrencauthon/SpecFlow/commit/3bdf54402c05b51e09be555fd1c715360185db6e

https://github.com/darrencauthon/SpecFlow/commit/8ee0b96c708da6dea6bae8aa8a73563c211b3cd6

Though I'm a code-testing, TDD maniac, I'm not intentionaly looking for untested code.  I spent most of my time reviewing the ```Queue```, examining around how the values are dequeued and trying out another way (which failed :stuck_out_tongue_winking_eye:).  There are just certain types of code that are harder to test, so I usually try poking around them to verify the tests were there.  If statements with one line of code, if statements with multiple conditions, string checks, one-off string replacements... ooh, ```IsNullOrEmpty```... guaranteed check for me.  :smile:

As someone who has worked in the Assist namespace, I want to say that the tests are worth it.  The code in the Assist namespace is not very complicated, but it is very detailed.  Remove one line here, case-insensitive GUID support goes away.  Remove one line there, ```sbyte``` support goes away.  Flip two lines, and table fields with spaces stop working.  Code that looks right can be very wrong.  Yet I've been able to make multiple major refactors to it over the years... all because it was tested to the hilt and most of it was TDD'd.

My suspicions went up with the two untested Table handlers... which I can just look at and know that they're probably not 100% in line with how tables are normally processed.  I'll help write some tests and get this stuff locked down and consistent with the rest of the library, but I think we should get these details worked out before merging.

samholder commented:
This is all great Darren. I'll also try and drive out where new tests are needed over the weekend, especially the enum part, it didn't occur to me to add tests for multiple arguments which are enums. I suppose its easy to assume that the tests are fairly comprehensive, but they might not be.

darrencauthon commented:
This PR, as well as the one I've been working on for step argument transformation support, inspired me to try this:

https://github.com/techtalk/SpecFlow/pull/459

I had trouble understanding the purpose of IBindingType.  Looking deeper into the code, I'm not sure it is needed.  Still working it out.

darrencauthon commented:
@samholder If you'd like, I'd be willing to help.  Adding tests to untested code is a specialty of mine. :smile:

samholder commented:
Get stuck in. I might not have time until later in the week anyway so any
help getting to a point where we can commit is all good.
On 10 Jul 2015 19:10, "Darren Cauthon" <notifications@github.com> wrote:

> @samholder <https://github.com/samholder> If you'd like, I'd be willing
> to help. Adding tests to untested code is a specialty of mine. [image:
> :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/techtalk/SpecFlow/pull/451#issuecomment-120482266>.
>

samholder commented:
@darrencauthon I've had no time to look at this, and I'm about to go on holiday for a month, so won't do any then. Apologies, life seems to have got in the way. I'll try and devote some time to it when I get back, but in the mean time we can either wait or roll back to the previous commit and merge that first.

I'm easy either way...

darrencauthon commented:
Sam, it’s no problem, I totally understand — things have been busy for me as well.  I meant to jump into it… just juggling.


We’ll get this done eventually. :)







Darren


—
Sent from Mailbox

On Thu, Jul 30, 2015 at 10:11 AM, Sam Holder <notifications@github.com>
wrote:

> @darrencauthon I've had no time to look at this, and I'm about to go on holiday for a month, so won't do any then. Apologies, life seems to have got in the way. I'll try and devote some time to it when I get back, but in the mean time we can either wait or roll back to the previous commit and merge that first.
> I'm easy either way...
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/techtalk/SpecFlow/pull/451#issuecomment-126366319

gasparnagy commented:
@samholder I just had a quick look on the PR and it looks good. I anyway hated that the converter was having both the "simple" and the complex conversion logic in one class. 
Have a good vacation and once you are back, I will also help finalizing this, especially considering my comments on #459 regarding the runtime types.

gasparnagy commented:
@samholder it seems i merged the pull requests in a wrong order so this is having a conflict now. can you please check it?

samholder commented:
Sure, I'll try and do it this weekend.

On Fri, Nov 27, 2015 at 10:56 AM, Gáspár Nagy <notifications@github.com>
wrote:

> @samholder <https://github.com/samholder> it seems i merged the pull
> requests in a wrong order so this is having a conflict now. can you please
> check it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/techtalk/SpecFlow/pull/451#issuecomment-160113868>.
>

samholder commented:
rebased on v2, and reverted the changes which tried to incorporate #397 as well as how that was meant to work was not clear, at least to me, which made testing difficult.

closing this PR
